add function to generate a free port in the range

ZiedZinelabidine opened the pull request:
As documented in https://github.com/pivotal-cf/cf-redis-release/issues/36, this PR enables control the shared instance port used within a preferred range.

- during configuration loading if ports are misconfigured it will report directly report the error. 
- When the service broker try to find a free port during provisioning and no port is available in the range it will give a service broker error. This error will ask to the user to contact his operator.

Regarding unit tests and integrations run as part of this PR: 
- integration tests were run with the docker image provided. I get some errors in this usecase, when a test want to check that program failed when a directory is not writeable it doesn't failed because on docker user privileged are root and this user is permitted to write on a folder which is not. Example: https://github.com/orange-cloudfoundry/cf-redis-broker/blob/shared-instance-ports-broker-prepation/brokerintegration/development_instance_provisioning_test.go#L127-L138
- all other tests look good.

Commit:
add function to generate a free port in the range

gberche-orange commented on system/free_tcp_port.go:
```diff
@@ -1,18 +1,33 @@
-	defer l.Close()
+func FindFreePort(num int) (int, error) {
+        t := strconv.Itoa(num)
+        toto := ":" + t
```
consider renaming variable toto into laddr as in the prototype of Net.Listen https://golang.org/pkg/net/#Listen

gberche-orange commented on system/free_tcp_port_test.go:
```diff
@@ -1,28 +1,29 @@
+        })
 
 })
+
```
Missing test case "returns an error if no free TCP port in the range"

gberche-orange commented on system/free_tcp_port_test.go:
```diff
@@ -1,28 +1,29 @@
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
+        . "github.com/onsi/ginkgo"
+        . "github.com/onsi/gomega"
```
Likely to have indentation issues. Please fix this, the diff is easier to read if only your changes show up.

gberche-orange reviewed (changes requested):
Please include the missing improvements, possibly with the help of @ArthurHlt

Commit:
add TU Creat Local_instance_creator_test.go

Commit:
fix errors

gberche-orange commented on system/free_tcp_port.go:
```diff
@@ -1,18 +1,33 @@
-	}
+func FindFreeInRangePort(Maxport int, Minport int)(int,error) {
+        i := Minport
+        port, err :=FindFreePort(i)
```
Please consider returning an error if the user mis configures:
- a max smaller or equal than the min parameter
- min or max smaller than 1024 or larger than 65535
& add a test case to cover it.

gberche-orange commented on system/free_tcp_port.go:
```diff
@@ -1,18 +1,33 @@
-		return -1, parseErr
-	}
+func FindFreeInRangePort(Maxport int, Minport int)(int,error) {
+        i := Minport
```
consider reordering parameter so that min is the 1st param, and max the 2nd param

gberche-orange commented on redis/local_instance_creator.go:
```diff
@@ -31,9 +31,10 @@ type LocalInstanceRepository interface {
 type LocalInstanceCreator struct {
 	LocalInstanceRepository
-	FindFreePort       func() (int, error)
+	FindFreePort       func(int) (int, error)
```
Reconsider removing the declaration in the struct which is not used anymore.

Commit:
add condition on FindFreeInRangPort parametres

Commit:
refactoring for test free tcp port

ArthurHlt commented:
All unit tests are OK. 
I did some refactoring to be able to test every use case

gberche-orange commented on system/free_tcp_port.go:
```diff
@@ -3,16 +3,50 @@ package system
+		return -1, errors.New("Not valid range port: maximum port is higher than " + strconv.Itoa(MIN_ACCEPTED_PORT))
+	}
+	port := minport
+	for port < maxport {
```
@ArthurHlt 
Minor case remaining to cover when minport=maxport, it should return a single port if available.

gberche-orange commented:
Thanks @ArthurHlt and @ZiedZinelabidine. Last changes look good to me (with the minor suggestion in https://github.com/orange-cloudfoundry/cf-redis-broker/pull/2#pullrequestreview-10120895 )

How do configuration errors or missing ports appear to bosh operators ? 
- would the bosh deployment nicely failing, and broker logs contain a trace with the error log ?
- would shared port misconfigurations only be detected at binding time, failing service binding ?

Please consider refining the PR text to illustrate how operators will discover such errors, and potentially propose/implement further enhancements to check config at broker start time.

Feel free then to merge (and squash commits) and submit pr against pivotal-cf/cf-redis-broker master (or possibly include minor

Commit:
Add test for free tcp port

Checking if the maximum port and minimum are equals that we should have this only port if available

Commit:
add port validation during configuration loading

ArthurHlt commented:
I've finally run integration tests with the docker image provided. I get some errors in this usecase, when a test want to check that program failed when a directory is not writeable it doesn't failed because on docker user privileged are root and this user is permitted to write on a folder which is not. Example: https://github.com/orange-cloudfoundry/cf-redis-broker/blob/shared-instance-ports-broker-prepation/brokerintegration/development_instance_provisioning_test.go#L127-L138

All other tests look good. 

About errors:
- during configuration loading if ports are misconfigured it will report directly report the error. 
- When the service broker try to find a free port during provisioning and no port is available in the range it will give a service broker error. This error will ask to the user to contact his operator.

gberche-orange commented:
Thanks @ArthurHlt for your contribution to unit tests and input validation in the configuration. Please merge and squash so that commits appear from your id.
